lib/Doctrine/ORM/Mapping/ReflectionReadonlyProperty.php

```diff
@@ -44,8 +45,19 @@ public function setValue(mixed $objectOrValue, mixed $value = null): void
 
         assert(is_object($objectOrValue));
 
-        if (parent::getValue($objectOrValue) !== $value) {
+        if ($this->isValueDifferent($objectOrValue, $value)) {
             throw new LogicException(sprintf('Attempting to change readonly property %s::$%s.', $this->class, $this->name));
         }
     }
+
+    private function isValueDifferent(object $object, mixed $value): bool
+    {
+        $objectValue = parent::getValue($object);
+
+        if (is_object($objectValue) && method_exists($objectValue, 'equals')) {
```

Comment:
I'm not comfortable with applying duck typing here. I understand the problem that you're trying to solve, but I'd rather not compare incomparable values than to guess how to compare them.

Reply:
I am not fan of duck typing here either, but other than adding explicit type checks (and thus requiring a non-dev dependency for every supported type), I am not sure how else the problem could be fixed

Reply:
I mean, we could just assume equality if we cannot compare the values. I would prefer that solution over calling a method that just looks like it does the right thing…

Reply:
Would exposing a configuration hook for providing custom equality checks be acceptable here?

Reply:
The above question still stands. I am happy to work towards another fix, but would like a confirmation that the approach is acceptable before I do so